### PR TITLE
Update dependencies and tighten TypeScript config

### PR DESCRIPTION
**Pull request by ivantha**
## Summary

Routine dependency updates and stricter TypeScript configuration.

## Dependency Changes

| Package | Old | New | Notes |
|---------|-----|-----|-------|
| express | 4.18 | 4.21 | Security patches |
| uuid | 9.x | 10.x | Major version bump |
| helmet | 7.x | 8.x | Major version bump |
| eslint | 8.x | 9.x | Major — flat config format |
| typescript | 5.3 | 5.6 | New features |
| @types/express | 4.x | 5.x | Major |
| @types/node | 20 | 22 | Match runtime |

## New Dependencies

- `express-rate-limit` — for #8
- `prettier` — code formatting

## TypeScript

- Target bumped to ES2022
- Added `noUnusedLocals`, `noUnusedParameters`, `noFallthroughCasesInSwitch`
- May surface new lint errors in existing code

## Risk

Major version bumps for `uuid`, `helmet`, `eslint`, and `@types/express` may require migration steps. Haven't tested all code paths yet.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,35 +1,42 @@
...
+    "morgan": "^1.10.0",
+    "express-rate-limit": "^7.4.0"
   },
   "devDependencies": {
-    "@types/express": "^4.17.21",
-    "@types/node": "^20.10.0",
-    "@types/uuid": "^9.0.7",
+    "@types/express": "^5.0.0",
+    "@types/node": "^22.0.0",
+    "@types/uuid": "^10.0.0",
```
Comment (ivantha) [inline package.json:28]: **[HIGH]** Major version bump for ESLint (8.x to 9.x) introduces breaking changes to configuration format.

The ESLint version has been updated from 8.x to 9.x. ESLint v9 introduces a new flat configuration file format (`eslint.config.js`) which is incompatible with the previous `.eslintrc.*` files. This is a significant breaking change that will require users to migrate their ESLint configuration. The `lint` script in `package.json` will likely fail until this migration is done.

- **Previous behavior**: ESLint used `.eslintrc.*` configuration files and a hierarchical loading system.
- **New behavior**: ESLint v9 defaults to a flat configuration file (`eslint.config.js`) and a different structure.
- **Who is affected**: Anyone running the `lint` script, and developers who maintain the ESLint configuration.
- **Migration path**: The project needs to create an `eslint.config.js` file and migrate all rules and settings from the existing ESLint configuration to the new format.

**Suggestion:**
```
Create an `eslint.config.js` file and migrate the ESLint configuration from `.eslintrc.*` to the new flat format. Update the `lint` script if necessary to point to the new configuration or ensure it's picked up automatically. Example migration steps can be found in the ESLint documentation.
```

File: package.json [hunk 2]
```diff
@@ -1,35 +1,42 @@
...
+    "test": "jest",
+    "format": "prettier --write src/",
+    "format:check": "prettier --check src/"
   },
   "dependencies": {
-    "express": "^4.18.2",
-    "uuid": "^9.0.0",
+    "express": "^4.21.0",
+    "uuid": "^10.0.0",
     "bcrypt": "^5.1.1",
```
Comment (ivantha) [inline package.json:18]: **[MEDIUM]** Major version bump for `uuid` (9.x to 10.x).

The `uuid` package has been updated from version 9.x to 10.x. While `uuid` v10 is largely backward-compatible, there might be subtle behavioral changes or deprecations. It's good practice to review the changelog for any potential impacts.

- **Previous behavior**: Used `uuid` v9.
- **New behavior**: Uses `uuid` v10.
- **Who is affected**: Any code that directly imports and uses the `uuid` library.
- **Migration path**: Review the `uuid` v10 changelog for any breaking changes or deprecations. Ensure existing usages are still valid.

**Suggestion:**
```
Review the release notes for `uuid` v10 to confirm backward compatibility and identify any potential migration steps. Add a comment in `package.json` or the PR description linking to the relevant release notes.
```

File: package.json [hunk 3]
```diff
@@ -1,35 +1,42 @@
...
     "jsonwebtoken": "^9.0.2",
     "cors": "^2.8.5",
-    "helmet": "^7.1.0",
-    "morgan": "^1.10.0"
+    "helmet": "^8.0.0",
```
Comment (ivantha) [inline package.json:21]: **[MEDIUM]** Major version bump for `helmet` (7.x to 8.x).

The `helmet` package has been updated from version 7.x to 8.x. Major version bumps for security middleware like Helmet often indicate changes in which headers are set or how they are configured. This could potentially affect how security headers are handled by the application.

- **Previous behavior**: `helmet` v7 was used.
- **New behavior**: `helmet` v8 is used.
- **Who is affected**: The application's security posture and potentially client-side interactions if certain headers change behavior.
- **Migration path**: Review the `helmet` v8 release notes and changelog for any breaking changes in header configuration or defaults. Ensure the application's security requirements are still met.

**Suggestion:**
```
Consult the `helmet` v8 release notes and changelog to understand any changes in default headers or configuration options. Verify that the application's security posture remains adequate after the update.
```

File: package.json [hunk 4]
```diff
@@ -1,35 +1,42 @@
...
+    "morgan": "^1.10.0",
+    "express-rate-limit": "^7.4.0"
   },
```
Comment (ivantha) [inline package.json:24]: **[MEDIUM]** Major version bump for `@types/express` (4.x to 5.x).

The `@types/express` package has been updated from version 4.x to 5.x. This major version bump usually means significant changes in the Express type definitions. It could affect how Express types are used throughout the codebase, potentially requiring adjustments to type annotations or imports.

- **Previous behavior**: TypeScript types for Express v4 were used.
- **New behavior**: TypeScript types for Express v5 are used.
- **Who is affected**: Any TypeScript code that interacts with Express types (e.g., `Request`, `Response`, `NextFunction`).
- **Migration path**: Review the `@types/express` v5 release notes and changelog. Adjust any TypeScript code that might be incompatible with the new type definitions.

**Suggestion:**
```
Check the release notes for `@types/express` v5. Make necessary adjustments to TypeScript code that uses Express types to ensure compatibility.
```

File: tsconfig.json [hunk 5]
```diff
@@ -1,8 +1,8 @@
 {
   "compilerOptions": {
-    "target": "ES2020",
+    "target": "ES2022",
     "module": "commonjs",
-    "lib": ["ES2020"],
+    "lib": ["ES2022"],
```
Comment (ivantha) [inline tsconfig.json:5]: **[HIGH]** TypeScript `target` bumped to `ES2022` without explicit `lib` updates, potentially causing issues.

The TypeScript `target` has been updated to `ES2022`. While the `lib` is also updated to `ES2022`, it's important to ensure that all necessary features supported by `ES2022` are available in the target runtime environment (Node.js v20.0.0 or higher, as per `engines` field). If the runtime environment doesn't fully support `ES2022` features, this could lead to runtime errors.

Additionally, the `lib` array should ideally include all necessary built-in types. Here it is updated correctly, but it's worth noting.

- **Previous behavior**: Target was `ES2020`.
- **New behavior**: Target is `ES2022`.
- **Who is affected**: Developers running the code in environments that do not fully support `ES2022` features.
- **Migration path**: Verify that the Node.js version (>= 20.0.0) used for running the application fully supports the ECMAScript features targeted by `ES2022`. If not, consider a lower target or polyfilling.

**Suggestion:**
```
Confirm that the Node.js runtime environment (specified as `>=20.0.0` in `package.json`) fully supports all ECMAScript features included in the `ES2022` target. Add a note in the PR description or documentation about this runtime requirement if it's a new constraint.
```

File: tsconfig.json [hunk 6]
```diff
@@ -12,7 +12,10 @@
     "resolveJsonModule": true,
     "declaration": true,
     "declarationMap": true,
-    "sourceMap": true
+    "sourceMap": true,
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
```
Comment (ivantha) [inline tsconfig.json:17]: **[LOW]** New TypeScript options `noUnusedLocals` and `noUnusedParameters` added.

The `noUnusedLocals` and `noUnusedParameters` TypeScript compiler options have been enabled. These are good practices for catching dead code and improving maintainability. However, they might surface a large number of new warnings or errors in the existing codebase if not carefully managed, potentially slowing down the initial adoption of these strict rules.

**Suggestion:**
```
Be prepared to address new linting errors related to unused variables and parameters. Consider enabling these flags gradually or addressing them incrementally if the initial volume of errors is overwhelming.
```

File: tsconfig.json [hunk 7]
```diff
@@ -12,7 +12,10 @@
...
+    "noFallthroughCasesInSwitch": true
```
Comment (ivantha) [inline tsconfig.json:18]: **[INFO]** New TypeScript option `noFallthroughCasesInSwitch` added.

The `noFallthroughCasesInSwitch` TypeScript compiler option has been enabled. This enforces the `break` statement after each `case` in a `switch` statement, which helps prevent accidental fall-through bugs. This is a good addition for code correctness.

**Suggestion:**
```
Ensure all `switch` statements correctly use `break` statements or comments indicating intentional fall-through (though the latter is discouraged with this flag).
```

File: package.json [hunk 8]
```diff
@@ -1,35 +1,42 @@
...
-    "uuid": "^9.0.0",
+    "express": "^4.21.0",
+    "uuid": "^10.0.0",
     "bcrypt": "^5.1.1",
     "jsonwebtoken": "^9.0.2",
     "cors": "^2.8.5",
-    "helmet": "^7.1.0",
-    "morgan": "^1.10.0"
+    "helmet": "^8.0.0",
+    "morgan": "^1.10.0",
```
Comment (ivantha) [inline package.json:22]: **[INFO]** New dependency `express-rate-limit` added.

The `express-rate-limit` middleware has been added as a new dependency. This is likely for rate limiting API requests, which is a good security and stability practice. No specific configuration changes are visible in the diff, so the default configuration is likely being used.

**Suggestion:**
```
Ensure that `express-rate-limit` is correctly configured (e.g., rate, windowMs, message) to prevent abuse and denial-of-service attacks. The current diff doesn't show configuration, so this might be handled elsewhere or defaults are relied upon.
```

**Review by ivantha — COMMENTED** [review]
## ◎ Odinro Review

The pull request updates dependencies, introduces new linting and formatting tools, and tightens the TypeScript configuration. Several major version bumps in dependencies, particularly ESLint, will require careful migration. The new TypeScript settings and added linting rules are positive steps towards code quality.

**Suggested Version Bump:** `major`

### ⚠️ Breaking Changes

- **[HIGH]** Major version bump for ESLint (8.x to 9.x) introduces breaking changes to configuration format.: The ESLint version has been updated from 8.x to 9.x. ESLint v9 introduces a new flat configuration file format (`eslint.config.js`) which is incompatible with the previous `.eslintrc.*` files. This is a significant breaking change that will require users to migrate their ESLint configuration. The `lint` script in `package.json` will likely fail until this migration is done.

- **Previous behavior**: ESLint used `.eslintrc.*` configuration files and a hierarchical loading system.
- **New behavior**: ESLint v9 defaults to a flat configuration file (`eslint.config.js`) and a different structure.
- **Who is affected**: Anyone running the `lint` script, and developers who maintain the ESLint configuration.
- **Migration path**: The project needs to create an `eslint.config.js` file and migrate all rules and settings from the existing ESLint configuration to the new format.
  > **Suggestion:** Create an `eslint.config.js` file and migrate the ESLint configuration from `.eslintrc.*` to the new flat format. Update the `lint` script if necessary to point to the new configuration or ensure it's picked up automatically. Example migration steps can be found in the ESLint documentation.
- **[MEDIUM]** Major version bump for `uuid` (9.x to 10.x).: The `uuid` package has been updated from version 9.x to 10.x. While `uuid` v10 is largely backward-compatible, there might be subtle behavioral changes or deprecations. It's good practice to review the changelog for any potential impacts.

- **Previous behavior**: Used `uuid` v9.
- **New behavior**: Uses `uuid` v10.
- **Who is affected**: Any code that directly imports and uses the `uuid` library.
- **Migration path**: Review the `uuid` v10 changelog for any breaking changes or deprecations. Ensure existing usages are still valid.
  > **Suggestion:** Review the release notes for `uuid` v10 to confirm backward compatibility and identify any potential migration steps. Add a comment in `package.json` or the PR description linking to the relevant release notes.
- **[MEDIUM]** Major version bump for `helmet` (7.x to 8.x).: The `helmet` package has been updated from version 7.x to 8.x. Major version bumps for security middleware like Helmet often indicate changes in which headers are set or how they are configured. This could potentially affect how security headers are handled by the application.

- **Previous behavior**: `helmet` v7 was used.
- **New behavior**: `helmet` v8 is used.
- **Who is affected**: The application's security posture and potentially client-side interactions if certain headers change behavior.
- **Migration path**: Review the `helmet` v8 release notes and changelog for any breaking changes in header configuration or defaults. Ensure the application's security requirements are still met.
  > **Suggestion:** Consult the `helmet` v8 release notes and changelog to understand any changes in default headers or configuration options. Verify that the application's security posture remains adequate after the update.
- **[MEDIUM]** Major version bump for `@types/express` (4.x to 5.x).: The `@types/express` package has been updated from version 4.x to 5.x. This major version bump usually means significant changes in the Express type definitions. It could affect how Express types are used throughout the codebase, potentially requiring adjustments to type annotations or imports.

- **Previous behavior**: TypeScript types for Express v4 were used.
- **New behavior**: TypeScript types for Express v5 are used.
- **Who is affected**: Any TypeScript code that interacts with Express types (e.g., `Request`, `Response`, `NextFunction`).
- **Migration path**: Review the `@types/express` v5 release notes and changelog. Adjust any TypeScript code that might be incompatible with the new type definitions.
  > **Suggestion:** Check the release notes for `@types/express` v5. Make necessary adjustments to TypeScript code that uses Express types to ensure compatibility.
- **[HIGH]** ESLint major version upgrade requires configuration migration.: The jump from ESLint 8.x to 9.x is a major breaking change due to the introduction of the flat configuration system. The current `lint` script will likely fail until the ESLint configuration is migrated from `.eslintrc.*` files to `eslint.config.js`.

- **Previous behavior**: ESLint configuration was managed via `.eslintrc.*` files.
- **New behavior**: ESLint v9 requires a `eslint.config.js` file for configuration.
- **Who is affected**: Developers running the `lint` script and anyone modifying ESLint configurations.
- **Migration path**: Create an `eslint.config.js` file and migrate all rules, plugins, and settings from the existing configuration.
  > **Suggestion:** Update the ESLint configuration to the new flat format (`eslint.config.js`). Consult the official ESLint migration guide for detailed steps.


### PR Description Issues

- **[INFO]** PR description is good but could be more detailed on testing.: The PR description provides a good overview of the dependency changes and TypeScript configuration updates. However, the statement 'Haven't tested all code paths yet' is a bit vague. A more detailed testing approach or confirmation of key functionalities tested would be beneficial.
  > **Suggestion:** Elaborate on the testing strategy. For example: 'Basic API endpoints (GET /tasks, POST /tasks) have been tested manually. Automated tests for core CRUD operations pass. Further testing of edge cases and specific middleware interactions is recommended.'

File: package.json [hunk 9]
```diff
@@ -1,35 +1,42 @@
...
+    "express-rate-limit": "^7.4.0"
   },
   "devDependencies": {
-    "@types/express": "^4.17.21",
-    "@types/node": "^20.10.0",
-    "@types/uuid": "^9.0.7",
+    "@types/express": "^5.0.0",
+    "@types/node": "^22.0.0",
+    "@types/uuid": "^10.0.0",
     "@types/bcrypt": "^5.0.2",
```
Comment (ivantha) [inline package.json:29]: **[INFO]** New dependency `prettier` added for code formatting.

The `prettier` package has been added as a dev dependency, along with `format` and `format:check` scripts in `package.json`. This is a positive step towards consistent code style across the project.

**Suggestion:**
```
Integrate Prettier into the development workflow, possibly via pre-commit hooks, to ensure consistent code formatting automatically.
```

File: package.json (hunk 1 above)
Comment (ivantha) [inline package.json:28]: **[HIGH]** ESLint major version upgrade requires configuration migration.

The jump from ESLint 8.x to 9.x is a major breaking change due to the introduction of the flat configuration system. The current `lint` script will likely fail until the ESLint configuration is migrated from `.eslintrc.*` files to `eslint.config.js`.

- **Previous behavior**: ESLint configuration was managed via `.eslintrc.*` files.
- **New behavior**: ESLint v9 requires a `eslint.config.js` file for configuration.
- **Who is affected**: Developers running the `lint` script and anyone modifying ESLint configurations.
- **Migration path**: Create an `eslint.config.js` file and migrate all rules, plugins, and settings from the existing configuration.

**Suggestion:**
```
Update the ESLint configuration to the new flat format (`eslint.config.js`). Consult the official ESLint migration guide for detailed steps.
```

File: package.json [hunk 10]
```diff
@@ -1,35 +1,42 @@
...
   "main": "dist/index.js",
   "scripts": {
     "build": "tsc",
     "start": "node dist/index.js",
     "dev": "ts-node src/index.ts",
     "lint": "eslint src/",
-    "test": "jest"
+    "test": "jest",
+    "format": "prettier --write src/",
+    "format:check": "prettier --check src/"
```
Comment (ivantha) [inline package.json:13]: **[HIGH]** Major version bump for `uuid`

The `uuid` package has been updated from `9.0.0` to `10.0.0`. Major version bumps can introduce breaking changes. According to the `uuid` changelog, version 10.0.0 includes changes to the `randomUUID()` function's behavior regarding entropy source and might have other API incompatibilities.

**Previous behavior**: Used `uuid` version `9.0.0`.
**New behavior**: Uses `uuid` version `10.0.0`.
**Who is affected**: Any code that directly uses the `uuid` library, especially for generating IDs.
**Migration path**: Review the `uuid` changelog for version 10.x and update any code that relies on specific behaviors or APIs that may have changed.

**Suggestion:**
```
Review the `uuid` changelog (https://github.com/uuidjs/uuid/releases/tag/v10.0.0) and update usage if necessary.
```

File: package.json [hunk 11]
```diff
@@ -1,35 +1,42 @@
...
   },
   "dependencies": {
```
Comment (ivantha) [inline package.json:15]: **[HIGH]** Major version bump for `helmet`

The `helmet` package has been updated from `7.1.0` to `8.0.0`. Major version bumps can introduce breaking changes. Version 8.0.0 of Helmet includes changes to default security headers and may require configuration adjustments.

**Previous behavior**: Used `helmet` version `7.1.0`.
**New behavior**: Uses `helmet` version `8.0.0`.
**Who is affected**: All parts of the application that use Helmet for setting security-related HTTP headers.
**Migration path**: Review the `helmet` changelog for version 8.x and adjust any custom configurations to match the new defaults or API.

**Suggestion:**
```
Consult the `helmet` changelog (https://github.com/helmetjs/helmet/releases/tag/v8.0.0) and update the application's Helmet configuration if needed.
```

File: package.json [hunk 12]
```diff
@@ -1,35 +1,42 @@
...
+    "helmet": "^8.0.0",
+    "morgan": "^1.10.0",
+    "express-rate-limit": "^7.4.0"
   },
   "devDependencies": {
-    "@types/express": "^4.17.21",
-    "@types/node": "^20.10.0",
-    "@types/uuid": "^9.0.7",
+    "@types/express": "^5.0.0",
+    "@types/node": "^22.0.0",
```
Comment (ivantha) [inline package.json:27]: **[HIGH]** Major version bump for `eslint` and adoption of flat config

The `eslint` package has been updated from `8.55.0` to `9.0.0`. This is a major version bump and introduces a significant change: the adoption of the "flat config" format. ESLint v9 nooriale.g., `eslint.config.js` instead of `.eslintrc.js`). This change is not reflected in the provided diff, but it's a major breaking change for the linting setup.

**Previous behavior**: Used ESLint with `.eslintrc.js` configuration.
**New behavior**: ESLint v9 uses the flat config format (e.g., `eslint.config.js`).
**Who is affected**: Developers running the `lint` script, and potentially the CI/CD pipeline.
**Migration path**: Create an `eslint.config.js` file and migrate the existing ESLint rules and configuration from `.eslintrc.js` to the new format.

**Suggestion:**
```
Create an `eslint.config.js` file and migrate the ESLint configuration. Consult the ESLint migration guide for v9.
```

File: package.json (hunk 3 above)
Comment (ivantha) [inline package.json:21]: **[HIGH]** Major version bump for `@types/express`

The `@types/express` package has been updated from `4.17.21` to `5.0.0`. This is a major version bump for the type definitions of Express. These changes can break code that relies on specific type signatures or existing Express functionality.

**Previous behavior**: Used `@types/express` version `4.17.21`.
**New behavior**: Uses `@types/express` version `5.0.0`.
**Who is affected**: All TypeScript code interacting with Express.
**Migration path**: Review the Express type definitions and update any code that might be affected by changes in type signatures or Express API availability.

**Suggestion:**
```
Review the changes in `@types/express` v5 and update TypeScript code accordingly. Check for any deprecated or removed Express features that are no longer reflected in the types.
```

File: tsconfig.json (hunk 5 above)
Comment (ivantha) [inline tsconfig.json:5]: **[MEDIUM]** Target bumped to ES2022 and new strict TypeScript flags

The TypeScript target has been updated from `ES2020` to `ES2022`, and the `lib` is updated accordingly. Additionally, strict flags `noUnusedLocals`, `noUnusedParameters`, and `noFallthroughCasesInSwitch` have been enabled. While these are generally good practices, they can introduce new errors in existing code if it doesn't strictly adhere to these rules or uses features that are no longer implicitly allowed.

**Previous behavior**: TypeScript target `ES2020`, `lib` `ES2020`, and less strict options.
**New behavior**: TypeScript target `ES2022`, `lib` `ES2022`, and enabled `noUnusedLocals`, `noUnusedParameters`, `noFallthroughCasesInSwitch`.
**Who is affected**: All TypeScript code in the project.
**Migration path**: Address any new TypeScript errors that appear after this change. This may involve cleaning up unused variables/parameters or ensuring all switch cases have explicit breaks.

**Suggestion:**
```
Thoroughly test the application after enabling these stricter TypeScript settings and address any resulting errors.
```

File: package.json [hunk 13]
```diff
@@ -1,35 +1,42 @@
...
   "description": "A simple task manager REST API",
   "main": "dist/index.js",
   "scripts": {
     "build": "tsc",
     "start": "node dist/index.js",
     "dev": "ts-node src/index.ts",
     "lint": "eslint src/",
-    "test": "jest"
+    "test": "jest",
+    "format": "prettier --write src/",
```
Comment (ivantha) [inline package.json:12]: **[MEDIUM]** Major version bump for `express` and `@types/node`

The `express` package is updated from `4.18.2` to `4.21.0`, and `@types/node` from `20.10.0` to `22.0.0`. While these are not explicitly called out as major breaking changes in the PR description's risk section, major version updates for core dependencies like Express and Node types should always be treated with caution. There's a possibility of subtle behavioral changes or deprecations.

**Previous behavior**: Used `express` v4.18.2 and `@types/node` v20.10.0.
**New behavior**: Uses `express` v4.21.0 and `@types/node` v22.0.0.
**Who is affected**: The entire application, as Express is the primary framework.
**Migration path**: Review the changelogs for Express `4.x` (4.19.0 to 4.21.0) and Node.js types `22.x` for any potential breaking changes or deprecated features.

**Suggestion:**
```
Review the changelogs for Express 4.19.0-4.21.0 and @types/node 22.0.0 for any potential breaking changes and update the codebase if necessary.
```

File: package.json [hunk 14]
```diff
@@ -1,35 +1,42 @@
...
+    "format:check": "prettier --check src/"
   },
   "dependencies": {
-    "express": "^4.18.2",
-    "uuid": "^9.0.0",
+    "express": "^4.21.0",
+    "uuid": "^10.0.0",
```
Comment (ivantha) [inline package.json:17]: **[INFO]** New dependency: `express-rate-limit`

A new dependency, `express-rate-limit`, has been added to the project. This is for implementing rate limiting on API requests.

**Impact**: This introduces new functionality and potential configuration points. Ensure it's configured appropriately to prevent abuse and isn't overly restrictive for legitimate users.

**Suggestion:**
```
Verify the configuration of `express-rate-limit` in the application code to ensure it meets security and usability requirements.
```

File: package.json (hunk 1 above)
Comment (ivantha) [inline package.json:28]: **[INFO]** New dependency: `prettier`

The `prettier` package has been added for code formatting. New scripts `format` and `format:check` have been introduced.

**Impact**: This promotes code consistency. Ensure the Prettier configuration aligns with team preferences.

**Suggestion:**
```
Review the default Prettier configuration and customize it if necessary. Integrate `prettier --check` into the CI pipeline for automated formatting checks.
```

File: package.json (hunk 12 above)
Comment (ivantha) [inline package.json:27]: **[INFO]** ESLint v9 migration required

The `eslint` dependency has been updated to v9.x, which mandates the use of the "flat config" format. The current diff does not show the creation or migration of an ESLint configuration file (e.g., from `.eslintrc.js` to `eslint.config.js`). This means the `lint` script will likely fail or run with an incorrect configuration until the migration is complete.

**Suggestion:**
```
Create an `eslint.config.js` file and migrate your existing ESLint rules. Refer to the ESLint documentation for detailed migration steps.
```

**Review by ivantha — COMMENTED** [review]
## ◎ Odinro Review

This PR updates dependencies and tightens the TypeScript configuration. It introduces several major version bumps for key libraries like `uuid`, `helmet`, `eslint`, and `@types/express`, which may require migration effort. The addition of new linting rules and a higher target for TypeScript compilation also increases the risk of surfacing new errors in existing code. A new dependency `express-rate-limit` and a formatting script are also added.

**Suggested Version Bump:** `major`

### ⚠️ Breaking Changes

- **[HIGH]** Major version bump for `uuid`: The `uuid` package has been updated from `9.0.0` to `10.0.0`. Major version bumps can introduce breaking changes. According to the `uuid` changelog, version 10.0.0 includes changes to the `randomUUID()` function's behavior regarding entropy source and might have other API incompatibilities.

**Previous behavior**: Used `uuid` version `9.0.0`.
**New behavior**: Uses `uuid` version `10.0.0`.
**Who is affected**: Any code that directly uses the `uuid` library, especially for generating IDs.
**Migration path**: Review the `uuid` changelog for version 10.x and update any code that relies on specific behaviors or APIs that may have changed.
  > **Suggestion:** Review the `uuid` changelog (https://github.com/uuidjs/uuid/releases/tag/v10.0.0) and update usage if necessary.
- **[HIGH]** Major version bump for `helmet`: The `helmet` package has been updated from `7.1.0` to `8.0.0`. Major version bumps can introduce breaking changes. Version 8.0.0 of Helmet includes changes to default security headers and may require configuration adjustments.

**Previous behavior**: Used `helmet` version `7.1.0`.
**New behavior**: Uses `helmet` version `8.0.0`.
**Who is affected**: All parts of the application that use Helmet for setting security-related HTTP headers.
**Migration path**: Review the `helmet` changelog for version 8.x and adjust any custom configurations to match the new defaults or API.
  > **Suggestion:** Consult the `helmet` changelog (https://github.com/helmetjs/helmet/releases/tag/v8.0.0) and update the application's Helmet configuration if needed.
- **[HIGH]** Major version bump for `eslint` and adoption of flat config: The `eslint` package has been updated from `8.55.0` to `9.0.0`. This is a major version bump and introduces a significant change: the adoption of the "flat config" format. ESLint v9 nooriale.g., `eslint.config.js` instead of `.eslintrc.js`). This change is not reflected in the provided diff, but it's a major breaking change for the linting setup.

**Previous behavior**: Used ESLint with `.eslintrc.js` configuration.
**New behavior**: ESLint v9 uses the flat config format (e.g., `eslint.config.js`).
**Who is affected**: Developers running the `lint` script, and potentially the CI/CD pipeline.
**Migration path**: Create an `eslint.config.js` file and migrate the existing ESLint rules and configuration from `.eslintrc.js` to the new format.
  > **Suggestion:** Create an `eslint.config.js` file and migrate the ESLint configuration. Consult the ESLint migration guide for v9.
- **[HIGH]** Major version bump for `@types/express`: The `@types/express` package has been updated from `4.17.21` to `5.0.0`. This is a major version bump for the type definitions of Express. These changes can break code that relies on specific type signatures or existing Express functionality.

**Previous behavior**: Used `@types/express` version `4.17.21`.
**New behavior**: Uses `@types/express` version `5.0.0`.
**Who is affected**: All TypeScript code interacting with Express.
**Migration path**: Review the Express type definitions and update any code that might be affected by changes in type signatures or Express API availability.
  > **Suggestion:** Review the changes in `@types/express` v5 and update TypeScript code accordingly. Check for any deprecated or removed Express features that are no longer reflected in the types.
- **[MEDIUM]** Target bumped to ES2022 and new strict TypeScript flags: The TypeScript target has been updated from `ES2020` to `ES2022`, and the `lib` is updated accordingly. Additionally, strict flags `noUnusedLocals`, `noUnusedParameters`, and `noFallthroughCasesInSwitch` have been enabled. While these are generally good practices, they can introduce new errors in existing code if it doesn't strictly adhere to these rules or uses features that are no longer implicitly allowed.

**Previous behavior**: TypeScript target `ES2020`, `lib` `ES2020`, and less strict options.
**New behavior**: TypeScript target `ES2022`, `lib` `ES2022`, and enabled `noUnusedLocals`, `noUnusedParameters`, `noFallthroughCasesInSwitch`.
**Who is affected**: All TypeScript code in the project.
**Migration path**: Address any new TypeScript errors that appear after this change. This may involve cleaning up unused variables/parameters or ensuring all switch cases have explicit breaks.
  > **Suggestion:** Thoroughly test the application after enabling these stricter TypeScript settings and address any resulting errors.


### PR Description Issues

- **[HIGH]** Missing migration steps for breaking changes: The PR description mentions 'Major version bumps for `uuid`, `helmet`, `eslint`, and `@types/express` may require migration steps' but does not provide any specific details or guidance on how to perform these migrations. This makes it harder for reviewers and future developers to understand the impact and necessary actions.

**Improvement**: For each major dependency update, briefly describe the nature of the breaking change and provide a link to the relevant changelog or migration guide.
  > **Suggestion:** For `uuid` v10, `helmet` v8, `eslint` v9, and `@types/express` v5, add a brief explanation of potential breaking changes and links to their respective migration guides or changelogs.
- **[MEDIUM]** Incomplete 'Risk' section: The 'Risk' section states 'Haven't tested all code paths yet.' While honesty is good, this implies incomplete testing for a PR with multiple major dependency updates. It would be better to outline the testing strategy or at least mention which areas were focused on during testing.

**Improvement**: Specify the testing approach taken (e.g., manual testing of core flows, running existing test suites) and highlight any areas that remain a concern.
  > **Suggestion:** Expand the 'Risk' section to detail the testing performed and any remaining areas of concern. For example: 'Ran existing Jest test suite, which passed. Manually verified core API endpoints (tasks, users, auth). Further testing of edge cases related to rate limiting is recommended.'
- **[LOW]** Missing motivation for adding `express-rate-limit`: The description mentions the addition of `express-rate-limit` for '#8' but does not provide context for this issue. The motivation behind adding rate limiting or the specific problem it solves is unclear from the description alone.

**Improvement**: Briefly explain why rate limiting is being implemented.
  > **Suggestion:** Add a sentence explaining the purpose of `express-rate-limit`, e.g., 'Added `express-rate-limit` to protect against brute-force attacks and ensure API stability, addressing issue #8.'
- **[LOW]** Enhance description of TypeScript changes: The description mentions updating the TypeScript target and adding new flags. While accurate, it could be more informative by briefly explaining the benefit of `ES2022` (e.g., access to newer ECMAScript features) and the specific purpose of the new flags (`noUnusedLocals`, `noUnusedParameters`, `noFallthroughCasesInSwitch`) in improving code quality.

**Improvement**: Add a short sentence explaining the benefits of the chosen TypeScript configurations.
  > **Suggestion:** For the TypeScript section, consider adding: 'Target bumped to ES2022 to leverage newer JavaScript features. Enabled `noUnusedLocals`, `noUnusedParameters`, and `noFallthroughCasesInSwitch` to enforce stricter code quality and prevent potential bugs.'
